drivers/spi/spi_ll_stm32.c

```diff
@@ -288,10 +288,27 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 		spi_context_update_tx(&data->ctx, 2, 1);
 	}
 
+	if (data->half_duplex) {
+		while (ll_func_spi_is_busy(spi)) {
+			/* NOP */
+		}
+
+		while (ll_func_rx_is_not_empty(spi)) {
+			/* Cleaning up RX fifo */
+			LL_SPI_ReceiveData8(spi);
+		}
+		/* Switching to RX */
+		LL_SPI_SetTransferDirection(spi, LL_SPI_HALF_DUPLEX_RX);
+	}
+
 	while (!ll_func_rx_is_not_empty(spi)) {
 		/* NOP */
 	}
 
+	if (data->half_duplex) {
+		LL_SPI_SetTransferDirection(spi, LL_SPI_HALF_DUPLEX_TX);
+	}
+
```

Comment:
Shouldn't this be done just before exiting the function  and after the RX ?

Reply:
We loop until the RX fifo is filled in, just above it. At this stage we can already switch back to TX, but I guess it won't change much if it moved at the end of the function, more readable actually.

Reply:
half-duplex seems to be very sensitive to timing: the time you switch from RX to TX, you might end up with garbage in the RX pipe. (basically as long as clock line is on, it will be "receiving").